### PR TITLE
add options --script, --buffer to sample exe

### PR DESCRIPTION
**Pull request by akrantz**
--buffer          Load script using JsCreateExternalArrayBuffer() and JsRun() instead of JsRunScript().
--script <file>   Specify an additional script file to load. Can be specified multiple times.

### CODE REVIEW

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 1]
```diff
@@ -53,6 +55,16 @@ class CommandLineArguments
                         this->port = std::stoi(std::wstring(argv[index]));
                     }
                 }
+                else if (!arg.compare(L"--script"))
+                {
+                    ++index;
```
Comment (kfarnung) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: Need to verify that there's another parameter so we don't read out of bounds.

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 2]
```diff
@@ -81,48 +93,65 @@ class CommandLineArguments
...
-//
-unsigned currentSourceContext = 0;
+
+std::wstring ConvertStringFromUtf8ToUtf16(const std::string& string)
+{
+    std::vector<wchar_t> contents(string.length());
+
+    if (MultiByteToWideChar(CP_UTF8, 0, string.c_str(), static_cast<int>(string.length()), contents.data(), static_cast<int>(contents.size())) == 0)
+    {
+        const char* message = "chakrahost: unable to convert string from UTF - 8 to UTF - 16.";
```
Comment (kfarnung) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: nit: UTF-8 and UTF-16

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 3]
```diff
@@ -131,15 +160,32 @@ JsErrorCode RunScript(const wchar_t* filename, JsValueRef* result)
...
+
+        // Load script from the file
+        std::string script = LoadScriptUtf8(fullPath);
+        if (script.empty())
+        {
+            return JsErrorInvalidArgument;
+        }
+
+        IfFailRet(JsPointerToString(fullPath, wcslen(fullPath), &sourceUrl));
+        IfFailRet(JsCreateExternalArrayBuffer((void*) script.c_str(), (unsigned int) script.length(), NULL, NULL, &scriptValueRef));
```
Comment (kfarnung) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: Can we avoid C-style casts?  The `(void*)` here is actually casting away const.  It's probably fine to do that, but I prefer when it's explicit.

File: bin/Debugger.Sample/Debugger.Sample.cpp (hunk 1 above)
Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Will fix. For both --port and --script, I reversed the expression to use `if (index < argc)` for better readability.

File: bin/Debugger.Sample/Debugger.Sample.cpp (hunk 2 above)
Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Will fix.

File: bin/Debugger.Sample/Debugger.Sample.cpp (hunk 3 above)
Comment (kfarnung) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: The lifetime here is suspect.  When you create an external array buffer it expects the backing data to be available via the buffer for the entire lifetime of the object.  In this case the backing data will be freed when `script` goes out of scope.  I think you'll need to create a class to own the data, create an instance on the heap, and provide a finalizer to delete the data when the array buffer is collected.

Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Is this what you want to see:  `reinterpret_cast<void*>(const_cast<char*>(script.c_str()))`?

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 4]
```diff
@@ -131,15 +163,32 @@ JsErrorCode RunScript(const wchar_t* filename, JsValueRef* result)
...
+
+        // Load script from the file
+        std::string script = LoadScriptUtf8(fullPath);
+        if (script.empty())
+        {
+            return JsErrorInvalidArgument;
+        }
+
+        IfFailRet(JsPointerToString(fullPath, wcslen(fullPath), &sourceUrl));
+        IfFailRet(JsCreateExternalArrayBuffer(const_cast<char*>(script.c_str()), static_cast<unsigned int>(script.length()), NULL, NULL, &scriptValueRef));
```
Comment (kfarnung) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: We still have the lifetime issue from my previous comment here.  We need to ensure that the data returned by `c_str` is valid for the entire lifetime of the external array buffer.

Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Yes, I pushed other changes but haven't had a chance to address this yet.

**Comment by akrantz** [comment]
@kfarnung I added a finalizer for the script buffer.

File: bin/Debugger.Sample/Debugger.Sample.cpp (hunk 4 above)
Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Fixed. Please see the latest commits.

**Comment by kfarnung** [comment]
Sorry for the delay, I was sick most of this week.  I did a quick pass and didn't see anything else, but I'll take another pass on Monday and land.

**Comment by kfarnung** [comment]
@akroshg Can you please take a look at these changes, test them out, and get this landed?

File: bin/Debugger.Sample/Debugger.Sample.cpp (hunk 3 above)
Comment (akrantz) [reply bin/Debugger.Sample/Debugger.Sample.cpp]: Fixed

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 5]
```diff
@@ -131,15 +189,34 @@ JsErrorCode RunScript(const wchar_t* filename, JsValueRef* result)
         return JsErrorInvalidArgument;
     }
 
-    // Load the script from the disk.
-    std::wstring script = LoadScript(fullPath);
-    if (script.empty())
+    if (loadScriptUsingBuffer)
     {
-        return JsErrorInvalidArgument;
+        JsValueRef sourceUrl;
```
Comment (akroshg) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: initialize them with JS_INVALID_REFERENCE

File: bin/Debugger.Sample/Debugger.Sample.cpp [hunk 6]
```diff
@@ -131,15 +189,34 @@ JsErrorCode RunScript(const wchar_t* filename, JsValueRef* result)
...
+        JsValueRef scriptValueRef;
```
Comment (akroshg) [inline bin/Debugger.Sample/Debugger.Sample.cpp]: (I don't know if my previous comment lost) - initialize this variable with JS_INVALID_REFERENCE

**Comment by akrantz** [comment]
@akroshg I've merged master and the last two commits make some minor fixes that I noticed.

**Pull request merged**